hera_cal/lstbin_simple.py

```diff
@@ -0,0 +1,1740 @@
+
+def adjust_lst_bin_edges(lst_bin_edges: np.ndarray) -> np.ndarray:
+    """
+    Adjust the LST bin edges so that they start in the range [0, 2pi) and increase.
+    
+    Performs the adjustment in-place.
+    """
+    while lst_bin_edges[0] < 0:
+        lst_bin_edges += 2*np.pi
+    while lst_bin_edges[0] >= 2*np.pi:
+        lst_bin_edges -= 2*np.pi
+
+def lst_bin_files_for_baselines(
+    data_files: list[Path | FastUVH5Meta], 
+    lst_bin_edges: np.ndarray, 
+    antpairs: Sequence[tuple[int, int]], 
+    freqs: np.ndarray | None = None, 
+    pols: np.ndarray | None = None,
+    cal_files: list[Path | None] | None = None,
+    time_arrays: list[np.ndarray] | None = None,
+    time_idx: list[np.ndarray] | None = None,
+    ignore_flags: bool = False,
+    rephase: bool = True,
+    antpos: dict[int, np.ndarray] | None = None,
+    lsts: np.ndarray | None = None,
+    redundantly_averaged: bool = False,
+    reds: RedundantGroups | None = None,
+    freq_min: float | None = None,
+    freq_max: float | None = None,
+) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, list[np.ndarray]]:
+    """Produce a set of LST-binned (but not averaged) data for a set of baselines.
+
+    This function takes a set of input data files, and reads any data in them that 
+    falls within the LST bins specified by ``lst_bin_edges`` (optionally calibrating
+    the data as it is read). The data is sorted into the LST-bins provided and returned
+    as a list of arrays, one for each LST bin. The data is not averaged within LST bins.
+
+    Only the list of baselines given will be read from each file, which makes it
+    possible to iterate over baseline chunks and call this function on each chunk,
+    to reduce maximum memory usage.
+
+    The data is binned via a simple histogram, i.e. the data represented at each LST
+    is essentially assumed to be a delta function in LST, and is fully assigned to one
+    particular LST bin. See :func:`lst_align` for details.
+    
+    Parameters
+    ----------
+    data_files
+        A list of paths to data files to read. Instead of paths, you can also pass
+        FastUVH5Meta objects, which will be used to read the data.
+    lst_bin_edges
+        A sequence of floats specifying the *edges* of the LST bins to use, with length
+        ``N_lstbins + 1``. Bins are thus assumed to be contiguous, but not necessarily
+        of equal size. 
+    antpairs
+        A list of antenna pairs to read from each file. Each pair should be a tuple
+        of antenna numbers. Note that having pairs in this list that are not present
+        in a particular file will not cause an error -- that file will simply not
+        contribute for that antpair.
+    freqs
+        Frequencies contained in the files. If not provided, will be read from the
+        first file in ``data_files``.
+    pols
+        Polarizations to read. If not provided, will be read from the first file in
+        ``data_files``.
```

Comment:
Yes, because we anyway have to write out a full UVData file, that has a data shape that is rectangular in (antpairs, pols). If we pull out specific 3-tuple baselines here, then we have to do work to spoof values that are in the full cross-product of antpairs-pols but don't exist in our 3-tuple baselines etc, and it can't be vectorized as well.